### ISSUE TITLE
Flutter freezes and crashes with surface assertion errors on the Web

### ISSUE THREAD

**Issue by Ahmadre** [flutter/flutter#50417]
## Steps to Reproduce

1. Run `flutter run -d chrome`.
2. Resize your browser window

**Expected results:**
Window updates itself and Gaps won't appear

**Actual results:**
<img width="1680" alt="Bildschirmfoto 2020-02-08 um 20 22 57" src="https://user-images.githubusercontent.com/18512224/74091106-27c4d080-4ab4-11ea-9461-66ee9b268950.png">

After chrome/firefox/safari/opera window resizes, flutter won't update and create white gaps in background. Sometimes it also freezes and crashes with the error:
<img width="835" alt="Bildschirmfoto 2020-02-08 um 20 24 49" src="https://user-images.githubusercontent.com/18512224/74091107-2eebde80-4ab4-11ea-9045-0d9c20f2adac.png">

Attention: The gap issue also happen on release builds but app won't crash.

<details>
  <summary>Logs</summary>

```
[  +26 ms] executing: [/Users/rebar/Documents/flutter/] git -c
log.showSignature=false log -n 1 --pretty=format:%H
[  +37 ms] Exit code 0 from: git -c log.showSignature=false log -n 1
--pretty=format:%H
[        ] 6dc3bfaa98e13c0f174e0cc229e7b43678d4d0c0
[        ] executing: [/Users/rebar/Documents/flutter/] git describe --match
v*.*.* --first-parent --long --tags
[  +22 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long
--tags
[        ] v1.15.2-37-g6dc3bfaa9
[   +9 ms] executing: [/Users/rebar/Documents/flutter/] git rev-parse
--abbrev-ref --symbolic @{u}
[   +6 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/master
[        ] executing: [/Users/rebar/Documents/flutter/] git ls-remote --get-url
origin
[   +8 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[  +53 ms] executing: [/Users/rebar/Documents/flutter/] git rev-parse
--abbrev-ref HEAD
[   +6 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] master
[   +4 ms] executing: sw_vers -productName
[  +16 ms] Exit code 0 from: sw_vers -productName
[        ] Mac OS X
[        ] executing: sw_vers -productVersion
[  +16 ms] Exit code 0 from: sw_vers -productVersion
[        ] 10.15.3
[        ] executing: sw_vers -buildVersion
[  +17 ms] Exit code 0 from: sw_vers -buildVersion
[        ] 19D76
[  +32 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping
update.
[   +5 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required,
skipping update.
[  +41 ms] executing: /Users/rebar/Library/Android/sdk/platform-tools/adb
devices -l
[   +8 ms] Exit code 0 from: /Users/rebar/Library/Android/sdk/platform-tools/adb
devices -l
[        ] List of devices attached
[   +9 ms] executing: /usr/bin/xcode-select --print-path
[  +13 ms] Exit code 0 from: /usr/bin/xcode-select --print-path
[        ] /Applications/Xcode.app/Contents/Developer
[   +1 ms] executing: /usr/bin/xcodebuild -version
[ +107 ms] Exit code 0 from: /usr/bin/xcodebuild -version
[   +1 ms] Xcode 11.3.1
           Build version 11C504
[   +2 ms] executing: xcrun --find xcdevice
[   +9 ms] Exit code 0 from: xcrun --find xcdevice
[        ] /Applications/Xcode.app/Contents/Developer/usr/bin/xcdevice
[   +2 ms] executing: xcrun xcdevice list --timeout 1
[+1534 ms] [
                      {
                        "simulator" : true,
                        "operatingSystemVersion" : "13.3 (17C45)",
                        "available" : true,
                        "platform" : "com.apple.platform.iphonesimulator",
                        "modelCode" : "iPhone10,5",
                        "identifier" : "830E878F-F0C5-4FFF-B0F4-62E25F9955E8",
                        "architecture" : "x86_64",
                        "modelName" : "iPhone 8 Plus",
                        "name" : "iPhone 8 Plus"
                      },
                      {
                        "simulator" : true,
                        "operatingSystemVersion" : "13.3 (17C45)",
                        "available" : true,
                        "platform" : "com.apple.platform.iphonesimulator",
                        "modelCode" : "iPad8,5",
                        "identifier" : "3678658B-C6F8-4EA5-B7F4-0EEE9326BBCE",
                        "architecture" : "x86_64",
                        "modelName" : "iPad Pro (12.9-inch) (3rd generation)",
                        "name" : "iPad Pro (12.9-inch) (3rd generation)"
                      },
                      {
                        "simulator" : true,
                        "operatingSystemVersion" : "13.3 (17C45)",
                        "available" : true,
                        "platform" : "com.apple.platform.iphonesimulator",
                        "modelCode" : "iPhone12,5",
                        "identifier" : "4C3E4855-3033-4F8B-BD55-ED013B547D7B",
                        "architecture" : "x86_64",
                        "modelName" : "iPhone 11 Pro Max",
                        "name" : "iPhone 11 Pro Max"
                      }
                    ]
[  +12 ms] /usr/bin/xcrun simctl list --json devices
[ +157 ms] Artifact Instance of 'AndroidMavenArtifacts' is not required,
skipping update.
[   +1 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping
update.
[   +5 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping
update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required,
skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required,
skipping update.
[ +310 ms] Generating
/Users/rebar/Documents/telexiom/Atemwegsliga_Web_neu/Hummingbird/android/app/src
/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
[ +171 ms] Launching lib/main.dart on Chrome in debug mode...
[        ] Building application for the web...
[+8429 ms] Starting daemon...
[ +200 ms] Initializing inputs
[  +17 ms] Reading cached asset graph...
[ +239 ms] Reading cached asset graph completed, took 236ms

[ +284 ms] Checking for updates since last build...
[ +677 ms] Checking for updates since last build completed, took 677ms

[  +13 ms] Initializing inputs
[   +2 ms] Reading cached asset graph...
[        ] Reading cached asset graph completed, took 236ms
[        ] Checking for updates since last build...
[        ] Checking for updates since last build completed, took 677ms
[+1759 ms] About to build [web]...
[   +4 ms] Running build...
[+1084 ms] 1.1s elapsed, 2/4 actions completed.
[+1094 ms] 2.2s elapsed, 2/4 actions completed.
[ +719 ms] Running build completed, took 2.8s

[        ] Caching finalized dependency graph...
[   +7 ms] Caching finalized dependency graph completed, took 114ms

[   +7 ms] Succeeded after 2.9s with 7 outputs (5 actions)

[  +11 ms] Building application for the web... (completed in 14,5s)
[        ] Attempting to connect to browser instance..
[+4121 ms] Debug service listening on ws://127.0.0.1:56383/ItspdSey2mU=

[ +413 ms]       Failed to load asset at path:
packages/build_web_compilers/src/dev_compiler/dart_sdk.js.

                          Status code: 404

                          Headers:
                          {
                      "date": "Sat, 08 Feb 2020 19:48:25 GMT",
                      "content-length": "9",
                      "x-frame-options": "SAMEORIGIN",
                      "content-type": "text/plain; charset=utf-8",
                      "x-xss-protection": "1; mode=block",
                      "x-content-type-options": "nosniff",
                      "server": "dart:io with Shelf",
                      "via": "1.1 shelf_proxy"
                    }

                          Content:
                          Not Found}

[  +30 ms] Attempting to connect to browser instance.. (completed in 4,6s)
[   +3 ms] Warning: Flutter's support for web development is not stable yet and
hasn't
[        ] been thoroughly tested in production environments.
[        ] For more information see https://flutter.dev/web
[   +1 ms] 🔥  To hot restart changes while running, press "r". To hot restart
(and refresh
           the browser), press "R".
[   +1 ms] For a more detailed help message, press "h". To quit, press "q".
[  +11 ms] Debug service listening on ws://127.0.0.1:56383/ItspdSey2mU=
```

```
Analyzing Hummingbird...                                                

   info • This class (or a class that this class inherits from) is marked as
          '@immutable', but one or more of its instance fields aren't final:
          CheckBoxListTileOption.value • lib/Components/DialogOption.dart:3:7 •
          must_be_immutable
   info • This function has a return type of 'String', but doesn't end with a
          return statement • lib/Components/Dialogs.dart:73:42 • missing_return
   info • This function has a return type of 'String', but doesn't end with a
          return statement • lib/Components/Dialogs.dart:108:46 • missing_return
   info • Unused import: 'package:atemwegsliga_web/Components/Dialogs.dart' •
          lib/Components/SideMenu.dart:1:8 • unused_import
   info • Unused import:
          'package:font_awesome_flutter/font_awesome_flutter.dart' •
          lib/Components/SideMenu.dart:13:8 • unused_import
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Components/SideMenu.dart:14:1 • avoid_web_libraries_in_flutter
   info • This class (or a class that this class inherits from) is marked as
          '@immutable', but one or more of its instance fields aren't final:
          SideMenu.user • lib/Components/SideMenu.dart:16:7 • must_be_immutable
   info • This class (or a class that this class inherits from) is marked as
          '@immutable', but one or more of its instance fields aren't final:
          LanguageTile._onLanguageTap, LanguageTile._title,
          LanguageTile.selected • lib/Containers/LanguageTile.dart:4:7 •
          must_be_immutable
   info • This class (or a class that this class inherits from) is marked as
          '@immutable', but one or more of its instance fields aren't final:
          DashedRect.borderRadius • lib/Containers/Shapes/Drawings.dart:36:7 •
          must_be_immutable
   info • Unused import:
          'package:atemwegsliga_web/Screens/Auth/Registration.dart' •
          lib/Containers/Shapes/RegistrationShapes.dart:2:8 • unused_import
   info • 'subhead' is deprecated and shouldn't be used. This is the term used
          in the 2014 version of material design. The modern term is subtitle1.
          This feature was deprecated after v1.13.8. •
          lib/Screens/Admin/Plugins/AnswersSearch.dart:155:51 •
          deprecated_member_use
   info • Unused import:
          'package:font_awesome_flutter/font_awesome_flutter.dart' •
          lib/Screens/Admin/Plugins/PatientSearch.dart:5:8 • unused_import
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Admin/Services/CSVExport.dart:2:1 •
          avoid_web_libraries_in_flutter
   info • Unused import:
          'package:font_awesome_flutter/font_awesome_flutter.dart' •
          lib/Screens/Admin/Sources/patientdata.source.dart:3:8 • unused_import
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Admin/index.dart:15:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Auth/Login.dart:17:1 • avoid_web_libraries_in_flutter
   info • This class (or a class that this class inherits from) is marked as
          '@immutable', but one or more of its instance fields aren't final:
          Login.email, Login.password • lib/Screens/Auth/Login.dart:19:7 •
          must_be_immutable
   info • The value of the field '_animation' isn't used •
          lib/Screens/Auth/Login.dart:34:13 • unused_field
   info • Unused import:
          'package:font_awesome_flutter/font_awesome_flutter.dart' •
          lib/Screens/Auth/Registration.dart:11:8 • unused_import
   info • The value of the field '_splashController' isn't used •
          lib/Screens/Auth/Registration.dart:28:28 • unused_field
   info • This function has a return type of 'String', but doesn't end with a
          return statement • lib/Screens/Auth/Registration.dart:215:24 •
          missing_return
   info • 'hasFloatingPlaceholder' is deprecated and shouldn't be used. Use
          floatingLabelBehaviour instead. This feature was deprecated after
          v1.13.2. • lib/Screens/Auth/Registration.dart:234:15 •
          deprecated_member_use
   info • This function has a return type of 'String', but doesn't end with a
          return statement • lib/Screens/Auth/Registration.dart:254:24 •
          missing_return
   info • 'hasFloatingPlaceholder' is deprecated and shouldn't be used. Use
          floatingLabelBehaviour instead. This feature was deprecated after
          v1.13.2. • lib/Screens/Auth/Registration.dart:272:15 •
          deprecated_member_use
   info • 'hasFloatingPlaceholder' is deprecated and shouldn't be used. Use
          floatingLabelBehaviour instead. This feature was deprecated after
          v1.13.2. • lib/Screens/Auth/Registration.dart:297:15 •
          deprecated_member_use
   info • 'hasFloatingPlaceholder' is deprecated and shouldn't be used. Use
          floatingLabelBehaviour instead. This feature was deprecated after
          v1.13.2. • lib/Screens/Auth/Registration.dart:315:15 •
          deprecated_member_use
   info • This function has a return type of 'String', but doesn't end with a
          return statement • lib/Screens/Auth/Registration.dart:706:28 •
          missing_return
   info • 'hasFloatingPlaceholder' is deprecated and shouldn't be used. Use
          floatingLabelBehaviour instead. This feature was deprecated after
          v1.13.2. • lib/Screens/Auth/Registration.dart:728:21 •
          deprecated_member_use
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Dashboard/Dashboard.dart:28:1 •
          avoid_web_libraries_in_flutter
   info • The value of the field '_isLoadingAnswers' isn't used •
          lib/Screens/Dashboard/Dashboard.dart:55:8 • unused_field
   info • This class (or a class that this class inherits from) is marked as
          '@immutable', but one or more of its instance fields aren't final:
          DetailsPage.answerSet •
          lib/Screens/Dashboard/Details/Details.dart:10:7 • must_be_immutable
   info • 'subhead' is deprecated and shouldn't be used. This is the term used
          in the 2014 version of material design. The modern term is subtitle1.
          This feature was deprecated after v1.13.8. •
          lib/Screens/Dashboard/Plugins/Search.dart:155:51 •
          deprecated_member_use
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Index.dart:12:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/PrivacyPolicy/PrivacyPolicy.dart:5:1 •
          avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Profile/Profile.dart:14:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Profile/Therapies/Therapies.dart:11:1 •
          avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Survey/AddSurvey/AddSurvey.dart:2:1 •
          avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Survey/AddSurvey/AddSurvey.dart:15:1 •
          avoid_web_libraries_in_flutter
   info • Don't explicitly initialize variables to null •
          lib/Screens/Survey/AddSurvey/AddSurvey.dart:23:14 • avoid_init_to_null
   info • 'await' applied to 'List<Map<String, dynamic>>', which is not a
          'Future' • lib/Screens/Survey/AddSurvey/AddSurvey.dart:140:41 •
          await_only_futures
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Tour/Tour.dart:12:1 • avoid_web_libraries_in_flutter
   info • Unused import:
          'package:font_awesome_flutter/font_awesome_flutter.dart' •
          lib/Screens/Wiki/Index.dart:7:8 • unused_import
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Screens/Wiki/Index.dart:8:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Services/answers.service.dart:1:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Services/lang.service.dart:3:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/Services/survey.service.dart:2:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/api/survey.dart:6:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/api/users.dart:6:1 • avoid_web_libraries_in_flutter
   info • Avoid using web-only libraries outside Flutter web plugin packages •
          lib/main.dart:17:1 • avoid_web_libraries_in_flutter
   info • 'Typography' is deprecated and shouldn't be used. The default
          Typography constructor defaults to the 2014 material design defaults.
          Applications are urged to migrate to Typography.material2018(), or, if
          the 2014 defaults are desired, to explicitly request them using
          Typography.material2014(). This feature was deprecated after v1.13.8.
          • lib/main.dart:115:35 • deprecated_member_use
   info • 'title' is deprecated and shouldn't be used. This is the term used in
          the 2014 version of material design. The modern term is headline6.
          This feature was deprecated after v1.13.8. • lib/main.dart:116:38 •
          deprecated_member_use
   info • This function has a return type of 'Locale', but doesn't end with a
          return statement • lib/main.dart:124:17 • missing_return
   info • 'display1' is deprecated and shouldn't be used. This is the term used
          in the 2014 version of material design. The modern term is headline4.
          This feature was deprecated after v1.13.8. • lib/main.dart:279:50 •
          deprecated_member_use
```

```
[✓] Flutter (Channel master, v1.15.3-pre.37, on Mac OS X 10.15.3 19D76, locale
    de-DE)
    • Flutter version 1.15.3-pre.37 at /Users/rebar/Documents/flutter
    • Framework revision 6dc3bfaa98 (17 hours ago), 2020-02-07 18:28:02 -0800
    • Engine revision 6158f03ef5
    • Dart version 2.8.0 (build 2.8.0-dev.8.0 514a8d4c84)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/rebar/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling
      support)
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.8.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 41.1.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.42.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.8.1

[✓] Connected device (2 available)
    • Chrome     • chrome     • web-javascript • Google Chrome 79.0.3945.130
    • Web Server • web-server • web-javascript • Flutter Tools

• No issues found!
```

</details>

**Comment by iapicca** [comment]
Hi @Ahmadre 
if possible could you please provide a reproducible minimal code sample?
Thank you

**Comment by Ahmadre** [comment]
@iapacca Is has nothing to do with specific code of mine. You can try it yourself just by:

1. clone flutter
2. switch to master channel
3. run flutter run -d chrome

and then watch the bug yourself. It happens on every flutter project.

Also run it on release builds. Here's a release example:
https://pulmonary-intensive-care-unit.firebaseapp.com/#/

**Comment by Ahmadre** [comment]
cc/ @iapicca same error as here: https://github.com/flutter/flutter/issues/50327

**Comment by ferhatb** [comment]
/cc @yjbanov : Did anything change in retained layer logic?
Stack trace has: assert(retainedSurface.isActive || retainedSurface.isReleased);

**Comment by yjbanov** [comment]
Not recently, no. Perhaps this exception is masking some other problem. BTW, this is not a Skia problem. This is in our HTML backend. I'll fix the labels.

**Comment by yjbanov** [comment]
So the error indicates that there's an attempt to add a layer as a retained layer that's neither active nor released, i.e. it is in one of `created` (new layer), `pendingRetention` (is scheduled to be retained),  or `pendingUpdate` (is scheduled to be updated). None of these states are valid for a layer that's being passed to `addRetained` (it either already did or never will). So I'll need to investigate how a layer got to one of these states in the first place. I will add more instrumentation on layer states.

@Ahmadre can you please provide a small code sample that reproduces this?

**Comment by Ahmadre** [comment]
@yjbanov sorry for the long delay. I started recognizing the problem when I used the navigation_rail package, but it's not the package but maybe the package including a pageView inside it.

> Also run it on release builds. Here's a release example:
> https://pulmonary-intensive-care-unit.firebaseapp.com/#/

Here's my dartpad minimal example: https://dartpad.dev/c0894a8d7d68fd0ec2daa2caa2879194

Just resize the view and watch the gaps and the cut-offs (when shrinking)

**Comment by rupertbulquerin** [comment]
facing the same issue on my button with gradient.. but when I remove the gradient the app works!

**Comment by Ahmadre** [comment]
Problem solved in DEV Channel v1.15.3

**Comment by ferhatb** [comment]
@ Ahmadre , thx for verifying fix.

**Comment by Ahmadre** [comment]
@ferhatb also solved in Master Channel (v1.15.4-pre.241) :) thx Flutter Team!
